### PR TITLE
fix(Scripts/IcecrownCitadel): Make Gunship Battle insta-wipe when no-one is outside of cannons

### PR DESCRIPTION
**Pull request by 0x7FFF**
Instant lose on Gunship Battle if no-one is outside

## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [X] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- None in the issues tab

## SOURCE:

Prior to Mists of Pandaria's patch 5.1 soloing Gunship Battle was not possible, because it would insta-wipe (insta-destroy) the ship.
https://www.bluetracker.gg/wow/topic/gb-en/5967569061-you-can-solo-gunship-cthun/

The changes have been validated through:
- [X] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [X] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. .go xyz -437.276 1959.755 208.174 631 (Horde)
.go xyz -437.257 2435.229 191.233 631 (Alliance)
2. Start Gunship Battle event
3. Sit inside cannon (solo)

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
build failing, also please update the branch to latest master

File: src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp [hunk 1]
```diff
@@ -673,6 +673,26 @@ class npc_gunship : public CreatureScript
...
 
+        void UpdateAI(uint32 diff) override
+        {
+            if (_instance->GetBossState(DATA_ICECROWN_GUNSHIP_BATTLE) != IN_PROGRESS)
+            {
+                return;
+            }
+
+            Map::PlayerList const& pl = me->GetMap()->GetPlayers();
+            for (Map::PlayerList::const_iterator itr = pl.begin(); itr != pl.end(); ++itr) {
```
Comment (Kitzunu) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp:684]: curly on a new line, see https://www.azerothcore.org/wiki/cpp-code-standards

File: src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp [hunk 2]
```diff
@@ -673,6 +673,26 @@ class npc_gunship : public CreatureScript
...
+                if (Player* p = itr->GetSource()) {
+                    if (!p->GetVehicle()) {
```
Comment (Kitzunu) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp:686]: same here

File: src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp (hunk 1 above)
Comment (Kitzunu) [reply src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp:684]: (or dont use it for oneliners)

**Review by Kitzunu — APPROVED** [review]

File: src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp [hunk 3]
```diff
@@ -673,6 +673,29 @@ class npc_gunship : public CreatureScript
...
         }
 
+        void UpdateAI(uint32 /*diff*/) override
+        {
+            if (_instance->GetBossState(DATA_ICECROWN_GUNSHIP_BATTLE) != IN_PROGRESS)
+            {
+                return;
+            }
+
+            Map::PlayerList const& pl = me->GetMap()->GetPlayers();
```
Comment (elthehablo) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_icecrown_gunship_battle.cpp:683]: Can use map->DoForAllPlayers here

**Comment by Nyeriah** [comment]
Question: at some point during the encounter every cannon gets frozen and players must kill the mages to resume the encounter. Therefore during that period of time, nobody will be inside a vehicle. I don't see this scenario being checked so is it safe to assume that a wipe will be triggered?

**Comment by Nyeriah** [comment]
Furthermore, please add this to a timer considering UpdateAI() ticks every millesecond and the current code iterates through the map player storage every millesecond which is not very resource efficient

**Comment by heyitsbench** [comment]
> I don't see this scenario being checked
```
if (!p->GetVehicle())
{
    // At least one player is outside a cannon, return --->without<--- resetting event.
    return;
}

**Comment by Nyeriah** [comment]
> > I don't see this scenario being checked
> 
> ```
> if (!p->GetVehicle())
> {
>     // At least one player is outside a cannon, return --->without<--- resetting event.
>     return;
> }
> ```

Thanks, I had misunderstood the logic behind it. The other point I've raised remains relevant though.

**Review by Kitzunu — CHANGES_REQUESTED** [review]
Please see to the reviews above

**Comment by qwe900** [comment]
i think with alot of players using autobalance or simimal things i would not change this.

**Comment by Kitzunu** [comment]
If this is blizzlike we should implement this in the core. Doesn't matter if it breaks a module, then the module author or maintainer can fix it from the module

**Comment by elthehablo** [comment]
Not to mention that it would be really easy to patch again in that module after fixing this

**Comment by LannyE** [comment]
Just to add my half cent to the discussion as a person who solos a lot, with the worldserver.conf being chock-full of non-blizzlike behavior to customize your experience would it be possible to make this as an option (defaulted to enable wipe, of course) in the conf?  This would make it so a server admin wouldn't have to wait for a fix to a module or module(s) to continue playing the way they used to.

**Comment by sudlud** [comment]
@0x7FFF what's the state of this PR? 🙂
There are pending reviews and also a suggestion to make it a config option to disable this feature maybe, do you want to update the PR?

**Comment by Kitzunu** [comment]
Closed in favor of https://github.com/azerothcore/azerothcore-wotlk/pull/19239

**Pull request closed without merging**